habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py

```diff
@@ -241,13 +274,169 @@ def add_receptacle_visualization(
+        """
+        Get all three vertices of a mesh triangle given it's face index as a list of numpy arrays.
+
+        :param f_ix: The index of the mesh triangle.
+        """
+        verts: List[np.ndarray] = []
+        for ix in range(3):
+            verts.append(
+                np.array(
+                    self.mesh_data[0][self.mesh_data[1][int(f_ix * 3 + ix)]]
+                )
+            )
+        return verts
+
+    def sample_area_weighted_triangle(self) -> int:
+        """
+        Isolates the area weighted triangle sampling code.
+
+        Returns a random triangle index sampled with area weighting.
+        """
+
+        def find_ge(a: List[Any], x) -> Any:
+            "Find leftmost item greater than or equal to x"
+            from bisect import bisect_left
+
+            i = bisect_left(a, x)
+            if i != len(a):
+                return i
+            raise ValueError
+
+        # first area weighted sampling of a triangle
+        sample_val = random.random()
+        tri_index = find_ge(self.area_weighted_accumulator, sample_val)
+        return tri_index
+
+    def sample_uniform_local(
+        self, sample_region_scale: float = 1.0
+    ) -> mn.Vector3:
+        """
+        Sample a uniform random point from the mesh.
+
+        :param sample_region_scale: defines a XZ scaling of the sample region around its center. For example to constrain object spawning toward the center of a receptacle.
+        """
+
+        if sample_region_scale != 1.0:
+            logger.warning(
+                "TriangleMeshReceptacle does not support 'sample_region_scale' != 1.0."
+            )
+
+        tri_index = self.sample_area_weighted_triangle()
+
+        # then sample a random point in the triangle
+        v = self.get_face_verts(f_ix=tri_index)
+        rand_point = random_triangle_point(v[0], v[1], v[2])
+
+        return rand_point
+
+    def debug_draw(
+        self, sim: habitat_sim.Simulator, color: Optional[mn.Color4] = None
+    ) -> None:
+        """
+        Render the Receptacle with DebugLineRender utility at the current frame.
+        Draws the Receptacle mesh.
+        Must be called after each frame is rendered, before querying the image data.
+
+        :param sim: Simulator must be provided.
+        :param color: Optionally provide wireframe color, otherwise magenta.
+        """
+        # draw all mesh triangles
+        if color is None:
+            color = mn.Color4.magenta()
+        dblr = sim.get_debug_line_render()
+        dblr.push_transform(self.get_global_transform(sim))
+        assert len(self.mesh_data[1]) % 3 == 0, "must be triangles"
+        for face in range(int(len(self.mesh_data[1]) / 3)):
+            verts = self.get_face_verts(f_ix=face)
+            for edge in range(3):
+                dblr.draw_transformed_line(
+                    verts[edge], verts[(edge + 1) % 3], color
+                )
+        dblr.pop_transform()
```

Comment:
Just FYI, the dedicated tool for drawing wireframes (without having to manually compose them out of line segments) is the [MeshVisualizer](https://doc.magnum.graphics/magnum/classMagnum_1_1Shaders_1_1MeshVisualizerGL3D.html) shader. I think it was used somewhere in the C++ viewer already, but no idea how easy is it to make it available here. 

In case the shader would be available, you'd keep the original MeshData (instead of taking just views on positions and indices from it) and draw it directly with the shader.

Reply:
Cool suggestion, I think this would be a cool way to do some debugging visualization in the future. Construct MeshData for the debug mesh and then register it for the MeshVisualizer shader. 
Is this exposed to python in some way that it could draw on top of an existing framebuffer or would it need to be built into the Habitat rendering pipeline on the C++ side?

Reply:
The shader isn't exposed at the moment (but I could do that easily), and yes that would work as long as you either have the framebuffer currently bound or have a way to access & bind it for drawing.

On the other hand, such direct approach would be _very_ incompatible with the batch renderer ... so for easier porting in the future it's probably best to just leave it as-is :sweat_smile: The debug line drawing will have to get ported to the batch renderer in any case, and I could then maybe expand it to handle also wireframe meshes somehow.